File: Regular Expression Matching/Regular Expression Matching.java
public class Solution {
    public boolean isMatch(String s, String p) {
        
        if(p.length() == 0){
            return s.length() == 0;
        }
        
        else if(p.length() == 1 || p.charAt(1) != '*'){
            if(!s.isEmpty() && (p.charAt(0) == s.charAt(0) || p.charAt(0) == '.')){
                return isMatch(s.substring(1), p.substring(1));
            }
            else{
                return false;
            }
        }
        
        else{
            // Discard consecutive '*' string
            int pIndex = 2;
            while(pIndex < p.length()){
                if(p.charAt(pIndex) != '*'){
                    break;
                }
                ++pIndex;
            }
            // Match s with p
            // Always try to match s with p.substring(pIndex)
            // If p.charAt(0) matches s.charAt(0)
            // try to match every substring of s with p.substring(pIndex)
            for(int i = 0; i <= s.length() && (i == 0 || s.charAt(i - 1) == p.charAt(0) || p.charAt(0) == '.'); ++i){
                if(isMatch(s.substring(i), p.substring(pIndex))){
                    return true;
                }
            }
            return false;
        }
        
    }
}

File: Regular Expression Matching/Regular Expression Matching.py
# Use dp to solve this problem
class Solution:
    # @return a boolean
    def isMatch(self, s, p):
        
        # dimension 1: p; dimension2: s
        dp = [[False for i in xrange(0, len(s) + 1)] for i in xrange(0, len(p) + 1)]
        dp[0][0] = True
        
        for i in xrange(1, len(p) + 1):
            if i >= 2 and p[i - 1] == '*':
                dp[i][0] = dp[i - 2][0]
                
        for i in xrange(1, len(s) + 1):
            for j in xrange(1, len(p) + 1):
                if p[j - 1] == '*':
                    # dp[j - 1][i]: originally match (w/o the asterisk)
                    # dp[j - 2][i]: match when p[i - 1] is discarded
                    # dp[j][i - 1]: match when p[j - 2] == s[i - 1]
                    dp[j][i] = dp[j - 1][i] or dp[j - 2][i] or (dp[j][i - 1] and (s[i - 1] == p[j - 2] or p[j - 2] == '.'))
                elif p[j - 1] == '.':
                    # matches one character
                    dp[j][i] = dp[j - 1][i - 1]
                else:
                    # s[i] == p[i]
                    dp[j][i] = (dp[j - 1][i - 1] and p[j - 1] == s[i - 1])
                    
        return dp[len(p)][len(s)]
